buidl/test/test_musig.py

```diff
@@ -0,0 +1,340 @@
+from itertools import combinations
+from random import randint
+from unittest import TestCase
+
+from buidl.ecc import N, G, S256Point
+from buidl.hd import HDPrivateKey
+from buidl.helper import SIGHASH_DEFAULT
+from buidl.taproot import MultiSigTapScript, MuSigTapScript, TapRootMultiSig
+from buidl.tx import Tx, TxIn, TxOut
+from buidl.witness import Witness
+
+
+class MuSigTest(TestCase):
+    def test_single_leaf_multisig(self):
+        hd_priv_key = HDPrivateKey.from_mnemonic(
+            "oil oil oil oil oil oil oil oil oil oil oil oil", network="signet"
+        )
+        # create a 3-of-5 single-leaf multisig
+        private_keys = [
+            hd_priv_key.get_p2tr_receiving_privkey(address=i) for i in range(5)
+        ]
+        points = [priv.point for priv in private_keys]
+        tr_multisig = TapRootMultiSig(points, 3)
+        tap_root_input = tr_multisig.single_leaf_tap_root()
+        tap_root_output = tr_multisig.multi_leaf_tap_root()
+        leaf = tap_root_input.tap_node
+        self.assertEqual(
+            tap_root_input.address(network="signet"),
+            "tb1p9gpzhc5fhlwlf49ze00fgjszxh5pl2p7az76758xwarweq08gcas8qa0r7",
+        )
+        prev_tx = bytes.fromhex(
+            "9baaec56eeef80bfa9c0f9c039fb71f134973179a3e2c3e8d037a4cc29cf6354"
+        )
+        tx_ins = []
+        for prev_index in range(10):
+            tx_in = TxIn(prev_tx, prev_index)
+            tx_in._value = 999130
+            tx_in._script_pubkey = tap_root_input.script_pubkey()
+            tx_ins.append(tx_in)
```

Comment:
I might be misunderstanding this, but from an API design perspective it's very confusing that each pubkey combinatorial is its own `TxIn`. In `buidl` a `TxIn` has always meant a specific UTXO.

Again, I could be misunderstanding as I attempt to port an example over to testnet.

Reply:
I'm trying to run this example on `testnet` for example and the script is failing. I think it's because I want `prev_index` to be both 1 (it is the 1th output index in [this transaction](https://blockstream.info/testnet/tx/e5dbd65d42e82bcd1cf08b9d767838de4dfc82eedfd04d104820d73407b64212)) and also an int of `range(10)`:
```
from buidl import *

hd_priv_key = HDPrivateKey.from_mnemonic(
    "oil oil oil oil oil oil oil oil oil oil oil oil", network="signet"
)
# create a 3-of-5 multi-leaf musig
private_keys = []
for i in range(5):
    private_keys.append(hd_priv_key.get_p2tr_receiving_privkey(address=i))
points = [priv.point for priv in private_keys]
tr_multisig = TapRootMultiSig(points, 3)
musig = MuSigTapScript(points)
tap_root = tr_multisig.musig_tap_root()
tap_root_input = tr_multisig.single_leaf_tap_root()
tap_root_output = tr_multisig.multi_leaf_tap_root()

assert tr_multisig.default_internal_pubkey == musig.point

# 2021-11-21 send testnet coins here:
assert tap_root.address(network="testnet") == "tb1pt8sygq8v5zjzrt03lwatrwm4vzvy3fktuhpv6hyhd63euw774vxq9u0vhn"

prev_tx = bytes.fromhex(
    "e5dbd65d42e82bcd1cf08b9d767838de4dfc82eedfd04d104820d73407b64212"
)
prev_index = 1

tx_ins = []
for prev_index, _ in enumerate(combinations(points, 3)):
    # range of 10 comes from all 3-of-5 combos)
    tx_in = TxIn(prev_tx, prev_index)
    tx_in._value = 998930
    tx_in._script_pubkey = tap_root_input.script_pubkey()
    tx_ins.append(tx_in)

# return to testnet faucet
tx_out = TxOut(tx_in.value() - 800, address_to_script_pubkey("mkHS9ne12qx9pS9VojpwU5xtRd4T7X7ZUt"))
tx_obj = Tx(1, tx_ins, [tx_out], 0, network="testnet", segwit=True)


for input_index, pubkeys in enumerate(combinations(points, 3)):
    tx_in = tx_ins[input_index]
    tx_in.witness.items = []
    leaf = MultiSigTapScript(pubkeys, 3).tap_leaf()
    # FIXME: this breaks here on the first loop
    assert tap_root_input.control_block(leaf) is True
    
    tx_obj.initialize_p2tr_multisig(
        input_index, tap_root_input.control_block(leaf), leaf.tap_script
    )
    sigs = []
    for priv in private_keys:
        if priv.point in pubkeys:
            sig = tx_obj.get_sig_taproot(input_index, priv, ext_flag=1)
            sigs.append(sig)
    tx_obj.finalize_p2tr_multisig(input_index, sigs)
    assert tx_obj.verify_input(input_index) is True
    assert tx_obj.verify() is True

```

Reply:
First, you're using both `tap_root` and `tap_root_input`. Get rid of `tap_root_input` as that's your input. You're going to want to send to the multi leaf taproot address which is tb1p7e95u84xfjeuyndvpzauavkw59m2r9ekw28xxax3cftgzx3kza8q7ruv3u, not tb1pt8sygq8v5zjzrt03lwatrwm4vzvy3fktuhpv6hyhd63euw774vxq9u0vhn, which is the musig taproot address.

Second, the original tx in the PR from signet is here: https://mempool.space/signet/tx/e6d03b31353c5bfcd2033db247d95909ed583e0d69d3bcbca9ddf31e98c861e4

As you can see, it literally has 10 inputs so we can make sure the same address can be unlocked 10 different ways using taproot. What you want to do in your script is something like:

```
prev_index = 1

tx_ins = []
tx_in = TxIn(prev_tx, prev_index)
tx_ins.append(tx_in)

# return to testnet faucet
tx_out = TxOut(tx_in.value() - 800, address_to_script_pubkey("mkHS9ne12qx9pS9VojpwU5xtRd4T7X7ZUt"))
tx_obj = Tx(1, tx_ins, [tx_out], 0, network="testnet", segwit=True)

tx_in.witness.items = []
leaf = MultiSigTapScript(pubkeys[:3], 3).tap_leaf()
assert tap_root_input.control_block(leaf) is True
    
tx_obj.initialize_p2tr_multisig(
        0, tap_root_input.control_block(leaf), leaf.tap_script
)
sigs = []
for priv in private_keys:
    if priv.point in pubkeys:
        sig = tx_obj.get_sig_taproot(input_index, priv, ext_flag=1)
        sigs.append(sig)
tx_obj.finalize_p2tr_multisig(input_index, sigs)
assert tx_obj.verify_input(input_index) is True
assert tx_obj.verify() is True
```